File: bridge/stdio_bridge.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# ============================================================================
# SU MCP Stdio Bridge
# 独立 Ruby 脚本，作为 MCP 客户端 (Claude Desktop 等) 与 SketchUp 之间的桥梁
#
# 工作原理:
#   MCP 客户端 <--stdio (JSON-RPC)--> stdio_bridge.rb <--TCP--> SketchUp TCP Server
#
# 使用方法:
#   ruby stdio_bridge.rb [host] [port]
#
# Claude Desktop 配置 (claude_desktop_config.json):
#   {
#     "mcpServers": {
#       "sketchup": {
#         "command": "ruby",
#         "args": ["C:/path/to/stdio_bridge.rb"],
#         "env": { "SU_MCP_PORT": "9876" }
#       }
#     }
#   }
# ============================================================================

require 'socket'
require 'json'

module SuMcpBridge

  DEFAULT_HOST = '127.0.0.1'
  DEFAULT_PORT = 9876

  CONNECT_RETRY_INTERVAL = 2     # 重连间隔 (秒)
  CONNECT_MAX_RETRIES    = 30    # 最大重试次数
  READ_TIMEOUT           = 30    # TCP 读取超时 (秒)

  class StdioBridge

    def initialize(host, port)
      @host    = host
      @port    = port
      @socket  = nil
      @running = false
    end

    # 启动桥接
    def run
      @running = true
      log("SU MCP Bridge starting, connecting to #{@host}:#{@port}...")

      connect_with_retry

      log('Connected to SketchUp MCP Server')

      # 主循环: 从 stdin 读取，转发到 TCP，回写 stdout
      while @running
        process_stdin
      end
    rescue Interrupt
      log('Bridge interrupted')
    rescue => e
      log("Bridge error: #{e.message}")
      # 写一个 JSON-RPC 错误到 stdout 通知客户端
      error_response = {
        'jsonrpc' => '2.0',
        'id'      => nil,
        'error'   => {
          'code'    => -32603,
          'message' => "Bridge error: #{e.message}",
        },
      }
      $stdout.puts(JSON.generate(error_response))
      $stdout.flush
    ensure
      disconnect
    end

    private

    # 带重试的 TCP 连接
    def connect_with_retry
      retries = 0

      loop do
        begin
          @socket = TCPSocket.new(@host, @port)
          @socket.setsockopt(Socket::IPPROTO_TCP, Socket::TCP_NODELAY, 1)
          return
        rescue Errno::ECONNREFUSED, Errno::ECONNRESET, Errno::ETIMEDOUT => e
          retries += 1
          if retries >= CONNECT_MAX_RETRIES
            raise "Cannot connect to SketchUp MCP Server at #{@host}:#{@port} " \
                  "after #{retries} attempts. Is SketchUp running with the plugin started?"
          end
          log("Connection attempt #{retries}/#{CONNECT_MAX_RETRIES} failed: #{e.message}, " \
              "retrying in #{CONNECT_RETRY_INTERVAL}s...")
          sleep(CONNECT_RETRY_INTERVAL)
        end
      end
    end

    # 处理 stdin 输入
    def process_stdin
      # 阻塞读取一行
      line = $stdin.gets

      unless line
        # stdin 关闭 (客户端断开)
        @running = false
        return
      end

      line = line.strip
      return if line.empty?

      # 验证是否为合法 JSON
      begin
        JSON.parse(line)
      rescue JSON::ParserError => e
        log("Invalid JSON from stdin: #{e.message}")
        error_response = {
          'jsonrpc' => '2.0',
          'id'      => nil,
          'error'   => {
            'code'    => -32700,
            'message' => "Parse error: #{e.message}",
          },
        }
        $stdout.puts(JSON.generate(error_response))
        $stdout.flush
        return
      end

      # 转发到 TCP
      begin
        send_to_tcp(line)
        response = read_from_tcp
        if response
          $stdout.puts(response)
          $stdout.flush
        end
      rescue Errno::EPIPE, Errno::ECONNRESET, IOError => e
        log("TCP connection lost: #{e.message}, reconnecting...")
        reconnect
        # 重试一次
        begin
          send_to_tcp(line)
          response = read_from_tcp
          if response
            $stdout.puts(response)
            $stdout.flush
          end
        rescue => retry_err
          log("Retry failed: #{retry_err.message}")
        end
      end
    end

    # 发送到 TCP
    def send_to_tcp(message)
      @socket.puts(message)
      @socket.flush
    end

    # 从 TCP 读取响应
    def read_from_tcp
      ready = IO.select([@socket], nil, nil, READ_TIMEOUT)

      unless ready
        log('TCP read timeout')
        return nil
      end

      line = @socket.gets
      line&.strip
    end

    # 重新连接
    def reconnect
      disconnect
      connect_with_retry
      log('Reconnected to SketchUp')
    end

    # 断开连接
    def disconnect
      @socket&.close rescue nil
      @socket = nil
    end

    # 日志输出到 stderr (不影响 stdout 的 JSON-RPC 通信)
    def log(message)
      $stderr.puts("[SU-MCP-Bridge] #{Time.now.strftime('%H:%M:%S')} #{message}")
      $stderr.flush
    end

  end # class StdioBridge

end # module SuMcpBridge

# --- 主程序入口 ---
if __FILE__ == $PROGRAM_NAME
  host = ARGV[0] || ENV['SU_MCP_HOST'] || SuMcpBridge::DEFAULT_HOST
  port = (ARGV[1] || ENV['SU_MCP_PORT'] || SuMcpBridge::DEFAULT_PORT).to_i

  bridge = SuMcpBridge::StdioBridge.new(host, port)
  bridge.run
end

File: bridge/stdio_bridge.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
SU MCP Stdio Bridge (Python)
=========================================
Cross-platform MCP stdio bridge for SketchUp.
Works on Windows, macOS, Linux without extra dependencies.

Usage:
    python3 stdio_bridge.py [host] [port]

Claude Desktop config:
    {
      "mcpServers": {
        "sketchup": {
          "command": "python3",
          "args": ["/path/to/bridge/stdio_bridge.py"],
          "env": { "SU_MCP_PORT": "9876" }
        }
      }
    }
    Note: On Windows use "python" instead of "python3" if needed.
"""

from __future__ import annotations

import json
import os
import socket
import sys
import time
from typing import BinaryIO, Tuple

# ---------------------------------------------------------------------------
# Configuration
# ---------------------------------------------------------------------------
HOST           = os.environ.get('SU_MCP_HOST', '127.0.0.1')
PORT           = int(os.environ.get('SU_MCP_PORT', '9876'))
RETRY_INTERVAL = 2    # seconds between connection retries
MAX_RETRIES    = 30   # give up after this many attempts
READ_TIMEOUT   = 30   # seconds to wait for a TCP response


def _log(msg: str) -> None:
    """Write to stderr so it doesn't interfere with JSON-RPC on stdout."""
    print(f'[SU-MCP-Bridge] {msg}', file=sys.stderr, flush=True)


# ---------------------------------------------------------------------------
# TCP Connection
# ---------------------------------------------------------------------------

def _connect() -> socket.socket:
    """Connect to SketchUp TCP server with retries."""
    for attempt in range(1, MAX_RETRIES + 1):
        try:
            sock = socket.create_connection((HOST, PORT), timeout=5)
            sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
            sock.settimeout(READ_TIMEOUT)
            return sock
        except (ConnectionRefusedError, OSError, TimeoutError) as exc:
            if attempt >= MAX_RETRIES:
                raise RuntimeError(
                    f'Cannot connect to SketchUp MCP Server at {HOST}:{PORT} '
                    f'after {MAX_RETRIES} attempts.\n'
                    f'Make sure SketchUp is running and the MCP Server is started '
                    f'(Extensions > SU MCP Server > Start Server).'
                ) from exc
            _log(f'Attempt {attempt}/{MAX_RETRIES} failed: {exc}'
                 f', retrying in {RETRY_INTERVAL}s...')
            time.sleep(RETRY_INTERVAL)
    # 这一行逻辑上不可达（MAX_RETRIES 次后 raise），
    # 但显式 raise 让 Pyre2 能正确推断函数总能返回 socket 或抛出异常
    raise RuntimeError('Unreachable: connection loop exhausted')


# ---------------------------------------------------------------------------
# Binary stdio helpers (handle Windows line-ending issues)
# ---------------------------------------------------------------------------

def _get_binary_stdio() -> Tuple[BinaryIO, BinaryIO]:
    """Return binary stdin/stdout. On Windows, switch to binary mode."""
    if sys.platform == 'win32':
        import msvcrt
        msvcrt.setmode(sys.stdin.fileno(), os.O_BINARY)
        msvcrt.setmode(sys.stdout.fileno(), os.O_BINARY)
    return sys.stdin.buffer, sys.stdout.buffer


# ---------------------------------------------------------------------------
# Main bridge loop
# ---------------------------------------------------------------------------

def _is_request(line: bytes) -> bool:
    """
    判断是否为 JSON-RPC request（有 id 字段）还是 notification（无 id 或 id 为 null）。
    MCP 协议：request 需要响应，notification 不需要响应。
    """
    try:
        msg = json.loads(line)
        # 有 'id' 字段且不为 None → request，需等待响应
        return isinstance(msg, dict) and msg.get('id') is not None
    except Exception:
        # 解析失败时保守地等待响应
        return True


def _make_sock_file(sock: socket.socket) -> BinaryIO:
    """将 socket 包装为二进制文件对象，并标注返回类型供 Pyre2 推断。"""
    return sock.makefile('rwb', buffering=0)  # type: ignore[return-value]


def main() -> None:
    # Allow host/port as positional args for convenience
    if len(sys.argv) >= 2:
        global HOST
        HOST = sys.argv[1]
    if len(sys.argv) >= 3:
        global PORT
        PORT = int(sys.argv[2])

    _log(f'Starting bridge → {HOST}:{PORT}')

    sock = _connect()
    sock_file: BinaryIO = _make_sock_file(sock)
    _log('Connected to SketchUp MCP Server')

    stdin, stdout = _get_binary_stdio()

    while True:
        # Read one line from Claude (MCP client) via stdin
        try:
            raw = stdin.readline()
        except (KeyboardInterrupt, EOFError):
            break

        if not raw:          # stdin closed
            break

        line = raw.strip()
        if not line:
            continue

        # Determine if this message is a request (needs response) or notification (no response)
        needs_response = _is_request(line)

        # Forward to SketchUp TCP server
        try:
            sock_file.write(line + b'\n')
            sock_file.flush()

            if needs_response:
                # Request: block waiting for response
                response = sock_file.readline()
            else:
                # Notification: do NOT wait for a response
                # (MCP notifications/initialized, etc. have no response)
                decoded = line.decode('utf-8', errors='replace')
                _log(f'Notification sent (no response expected): {decoded[:80]}')  # type: ignore
                continue

        except (BrokenPipeError, ConnectionResetError, OSError) as exc:
            _log(f'TCP connection lost: {exc}, reconnecting...')
            try:
                sock.close()
            except Exception:
                pass
            sock = _connect()
            sock_file = _make_sock_file(sock)
            # Retry once after reconnect (only for requests)
            if needs_response:
                try:
                    sock_file.write(line + b'\n')
                    sock_file.flush()
                    response = sock_file.readline()
                except OSError as retry_exc:
                    _log(f'Retry failed: {retry_exc}')
                    continue
            else:
                continue

        if response:
            stdout.write(response.strip() + b'\n')
            stdout.flush()


if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        _log('Bridge stopped')
    except Exception as exc:
        _log(f'Fatal error: {exc}')
        sys.exit(1)
